### ISSUE TITLE
--footprints parameter

### ISSUE THREAD

**Issue by valgur** [sentinelsat/sentinelsat#153]
## Question
**Should we change the `--footprints` parameter's behavior to also include the name of the file that will be created?**

Let's make it a poll. I encourage anyone to leave a thumbs up or down to voice your opinion. Comments are welcome too, of course.

## Reasoning
I've been thinking that the `--footprints` flag would be more useful if it also took the name/path of the file to be created instead of always writing it to `search_footprints.geojson`.

An example use case might be:
```
sentinelsat -g germany_aoi.geojson --sentinel 1 --footprints s1_germany_footprints.geojson
```
On the other hand, forcing the user to always provide the destination filename is a minor inconvenience as well (which we could mitigate a bit by automatically appending the `.geojson` suffix if it is missing).

Since I rarely use the `--footprints` feature myself (I rely on the Python API) and the choice is not obvious, I would like to hear everyone else's take on this.

**Comment by Fernerkundung** [comment]
I've been thinking about this for some time now, as the `search_footprints.geojson` implementation was never supposed to be long term - alas I never came around to change it.

What I would expect from this function:

- `search_footprints.geojson` should be saved under `--path` (if given) otherwise at the current directory
- make it a keyword argument, instead of a flag, and use the filename given, otherwise fall back to `search_footprints.geojson`

That way we could get the best of both worlds - convenience and more options for the users to controll it. Added benefit is, that this change would be backwards compatible.

This would mean:

- `sentinelsat --footprints` results in `/<current dir>/search_footprints.geojson`
-  `sentinelsat --footprints s1_germany_footprints.geojson` results in `/<current dir>/s1_germany_footprints.geojson`
- `sentinelsat --footprints` results in `/<current dir>/search_footprints.geojson`
- `sentinelsat --path /some/other/dir --footprints s1_germany_footprints.geojson` results in `/some/other/dir/s1_germany_footprints.geojson`

**Comment by j08lue** [comment]
Can `click` do that at all - an option that works both as a flag (i.e. without following arguments) and as an option that takes a single argument? And try to explain that to the user...

I'd say make `--footprints` an option that **always requires** a single argument, which can also be `.` (very quick to type) and `if os.path.isdir(footprints_path)` join it with the default `search_footprints.geojson`.

**Comment by Fernerkundung** [comment]
> Can click do that at all - an option that works both as a flag (i.e. without following arguments) and as an option that takes a single argument?

You are right, implementing a flag that is also an option but only if the user provides input is indeed not supported by click (and might be hard to explain to the user).

>  make `--footprints` an option that always requires a single argument, which can also be . (very quick to type) and if os.path.isdir(footprints_path) join it with the default search_footprints.geojson

So you want users to provide essentially a `--(footprint)-path`? I'd rather we use `--path` for that and use `--footprint` and require the user to provide the filename and append the suffix, like @valgur initially suggested.

**Comment by j08lue** [comment]
I wonder how people are using `--footprints` - or should be using it.

I imagine you want to have the footprints when you download a limited set of files that you will keep in a designated folder, like `my_s2_data_for_project_A`. Then it makes a lot of sense to store the footprints file together with those files so you quickly can see the data coverage and find the right data file. And I guess then it does not matter so much what the file is called, since you have it in a unique directory and it will likely be the only `.geojson` file in that directory.

But it could also be that you want the footprints prior to downloading to evaluate the data coverage and decide which files to actually download. In that case, you might want to store them all in one directory under unique names like `footprints_L1C.geojson` and `footprints_L2A.geojson`. Then you indeed need to be able to specify the filename / full path.

So I guess the easiest is indeed to have the user specify a path and add the extension automatically, just like the initial suggestion.

**Comment by Fernerkundung** [comment]
Typical usage of the `--footprints` option I see at my workplace:

- exploring available data coverage (temporal and spatial) for a (new) study area
- saving the footprints as metadata in a folder together with the downloads to have a quick overview of the contents when the coverage is not large enough to warrant a real database (especially useful in environments with numerous small projects/study areas)

> have the user specify a path and add the extension automatically

A path or a filename? i.e. `--path /my_s2_data_for_project_A --footprints footprints_L1C` or `--footprints /my_s2_data_for_project_A/footprints_L1C`? I find the first one more intuitive, as the path option then is used to define where everything is stored (downloads and/or footprints).

**Comment by j08lue** [comment]
> A path or a filename?

Either: Just `os.path.join` it with `--path` so that the user can use relative paths (like just a filename) or absolute paths that can be outside of `--path`.

**Comment by jerryfrancis-97** [comment]
Hi @valgur,  i would like to try this issue. Can you please tell the entry point ?

**Comment by j08lue** [comment]
Nice, @jerryfrancis-97, please go for it. We only have one entrypoint (`sentinelsat`). 

Reading this now, I'd prefer the following solution:

1. `--footprints` always requires a single argument
2. `os.path.join` with `--path` (which defaults to `'.'`, so it always has a value).
3. No ambiguity about file or dir: You always need to pass a file name.

**Comment by jerryfrancis-97** [comment]
@j08lue ,  Are you expecting this sort of modification?

@click.option(
    "--footprints",
    is_flag=True,
    type=click.Path(exists=True),
    help="""Create a geojson file search_footprints.geojson with footprints
    and metadata of the returned products.
    """,
)

and..

    if footprints is True:
        footprints_geojson = api.to_geojson(products)
        if os.path.isdir(footprints):
            foot_path = os.path.join(footprints, "search_footprints.geojson")
        else:
            foot_path = "search_footprints.geojson"
        if path == ".":
            dump_path = os.path.join(os.getcwd(), foot_path)
        else:
            dump_path = os.path.join(path, foot_path) 
        with open(dump_path, "w") as outfile:
            outfile.write(gj.dumps(footprints_geojson))

i'm little confused...

**Comment by j08lue** [comment]
@jerryfrancis-97, no, this is not how `click` works. `--footprints` cannot be both a flag and a path. It should be a path only, and one that does not exist yet (duh, it is the path to the file to be created). Please make a PR and we will review it. 😄

**Comment by IpsumCapra** [comment]
Hey, I'm working on this issue, too, now. Hopefully you can expect a PR soon.

**Comment by valgur** [comment]
Implemented in #462.
